src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs

```diff
@@ -334,13 +339,20 @@ private bool HasDependencyToProcess
             }
         }
 
-        private HeadersSyncBatch BuildNewBatch()
+        private HeadersSyncBatch? BuildNewBatch()
         {
-            HeadersSyncBatch batch = new();
-            batch.MinNumber = _lowestRequestedHeaderNumber - 1;
-            batch.StartNumber = Math.Max(HeadersDestinationNumber, _lowestRequestedHeaderNumber - _headersRequestSize);
-            batch.RequestSize = (int)Math.Min(_lowestRequestedHeaderNumber - HeadersDestinationNumber, _headersRequestSize);
-            _lowestRequestedHeaderNumber = batch.StartNumber;
+            HeadersSyncBatch? batch = null;
+            do
+            {
+                batch = new();
+                batch.StartNumber = Math.Max(HeadersDestinationNumber,
+                    _lowestRequestedHeaderNumber - _headersRequestSize);
+                batch.RequestSize = (int)Math.Min(_lowestRequestedHeaderNumber - HeadersDestinationNumber,
+                    _headersRequestSize);
+                _lowestRequestedHeaderNumber = batch.StartNumber;
+
+                batch = ProcessPersistedPortion(batch);
+            } while (batch is null && ShouldBuildANewBatch());
```

Comment:
This is sligthly confusing, shouldn't the loop be outside of this method?

Reply:
Moved outside